### DISCUSSION TITLE
Use builtin libraries with Multiple dependent languages

### DISCUSSION THREAD

**Discussion by chbo123456** [eclipse-langium/langium#1363] [Q&A] [answered]
Hi,  I recently discovered Multiple dependent languages feature and made it work with a simplified definition-implementation example. Is the "builtin library" also working with this out of the box? I'd like to have a definition file loaded at plugin load, that would be the idea. I tried to simply integrate it like in given example (https://langium.org/guides/builtin-library/), but somehow this won't work. Do I have to adapt the shared service?

Thanks!

**Comment by msujew** [comment, marked as answer]
Hey @chbo123456,

Have you registered the library file under the correct file extension? Can you give a bit more context on what doesn't work?

We use the feature quite extensively in multi-language projects so we are certain that it should work as expected.

**Reply by chbo123456** [reply]
Hey @msujew , thanks for the quick response. 

Do you mean to register in the "createMyLanguageServices"?

Before answering, please allow me to rephrase my question/issue, as I changed some ideas in my mind:
Is it possible to load the definition part of the language on startup from a file that lies outside of the workspace (somewhere in the fs)? I initially thought this only possible with builtin libraries but maybe you have a better way.

**Reply by msujew** [reply]
> Is it possible to load the definition part of the language on startup from a file that lies outside of the workspace (somewhere in the fs)? I initially thought this only possible with builtin libraries but maybe you have a better way.

Yes, that's a possible way to use the workspace initialization phase. In the end, it's just a special phase of the language server that allows to load documents from outside the workspace before processing the workspace itself.

> Do you mean to register in the "createMyLanguageServices"?

There's a line to create a new document in the guide:

```ts
collector(this.documentFactory.fromString(builtinHelloWorld, URI.parse('builtin:///library.hello')));
```

That effectively registers the builtin library in the language server. The file extension (here `.hello`) is supposed to be the one used by your definition language.

**Reply by chbo123456** [reply]
Hey @msujew ,
just had time to look over the code again and, yeah, it was the wrong file extension together with some forgotten wip changes. Now it works perfectly. 

Btw, I simply read the definition file directly as a string into the "builtinHelloWorld" variable, that does the trick for "dynamic definition input" (but it seems that this can't be done asynchronously). Is there a way to easily notify the user with a note/popup on the ui, if the file does not exist?

Thanks, Christian

**Reply by cdietrich** [reply]
i guess you always can do 
`this.services.lsp.Connection?.window.showInformationMessage(...)`

**Reply by chbo123456** [reply]
@cdietrich sorry for the stupid question, but in which class can I do this? Doesn't seem to be part of the DefaultWorkspaceManager...

**Reply by cdietrich** [reply]
DefaultWorkspaceManager has shared services as constructor parameter. so simply assign it to property in your subclass override constructor

**Reply by msujew** [reply]
@chbo123456 You can set the connection yourself if you want:

```
private connection: Connection;
constructor(services: LangiumSharedServices) {
  this.connection = services.lsp.Connection;
}
```

And then use that value later.

**Reply by chbo123456** [reply]
Thanks for your help, guys!
